### PR TITLE
feat(perf): limit the number of request when a user is updated throug…

### PR DESCRIPTION
**Pull request by leleueri**
…h SCIM or authentication

fixes AM-685

## Test done (using Mongo and PSQL)

* Update user no email changes     :OK:
* Update user email changes        :OK:
* Update user roles change         :OK:
* Update user address change       :OK:
* Update user entitlement change   :OK:
* Update user photo change         :OK:
* Update user ims   change         :OK:
* Update user phoneNumbers change  :OK:

Ex update scim Request:

```
curl -X PUT -H"Authorization: Bearer ${TOKEN}" http://localhost:8092/updateuser/scim/Users/29db2ac2-3863-491d-9b2a-c23863491d75 -d'
{
  "schemas" : [ "urn:ietf:params:scim:schemas:core:2.0:User" ],
  "id" : "29db2ac2-3863-491d-9b2a-c23863491d75",
  "externalId" : "97fa302d-812a-4267-ba30-2d812a226776",
  "meta" : {
    "resourceType" : "User",
    "created" : "2023-07-12T11:29:19.905Z",
    "lastModified" : "2023-07-12T11:29:19.905Z",
    "location" : "http://localhost:8092/updateuser/scim/Users/29db2ac2-3863-491d-9b2a-c23863491d75"
  },
  "userName" : "user01",
  "name" : {
    "familyName" : "Leleu",
    "givenName" : "Eric"
  },
  "displayName" : "user01 user",
  "active" : true,
  "emails" : [ {
    "value" : "ele@acme.fr",
    "primary" : true
  },{
    "value" : "ele00000@acme.fr",
    "primary" : false
  } ],
  "roles" : [ "7f75ccd7-60ee-4623-b5cc-d760ee262342", "211181b3-dbce-4855-9181-b3dbce68551c" ],
  "addresses": [
  	{
  		"type":"office",
  		"region":"nord"
  	}
  ],
  "entitlements": ["reteter"],
  "phoneNumbers" : [ {
    "value" : "0615492508",
    "primary" : true
  }],
  "ims" : [ {
    "value" : "0ims",
    "primary" : true
  }],
  "photos" : [ {
    "value" : "photo492508",
    "primary" : true
  }]
}'
```

* reset password                   :OK:
* login                            :OK:
* enrich user profile              :OK:

* Patch user phoneNumbers add      :OK:
* Patch user phoneNumbers replace  :OK:
* Patch user phoneNumbers remove   :OK:

Ex Pacth scim request :

```
curl -X PATCH -H"Authorization: Bearer ${TOKEN}" http://localhost:8092/updateuser/scim/Users/29db2ac2-3863-491d-9b2a-c23863491d75 -d'
   { "schemas":
       ["urn:ietf:params:scim:api:messages:2.0:PatchOp"],
     "Operations":[
       {
        "op":"add",
        "path":"phoneNumbers",
        "value":[
         {
    "value" : "0615002508",
    "primary" : false
         }
        ]
       }
     ]
   }'
```

```
curl -X PATCH -H"Authorization: Bearer ${TOKEN}" http://localhost:8092/updateuser/scim/Users/29db2ac2-3863-491d-9b2a-c23863491d75 -d'
   { "schemas":
       ["urn:ietf:params:scim:api:messages:2.0:PatchOp"],
     "Operations":[
       {
        "op":"replace",
        "path":"phoneNumbers",
        "value":[
         {
    "value" : "0615002508",
    "primary" : true
         }
        ]
       }
     ]
   }'
```

```
curl -X PATCH -H"Authorization: Bearer ${TOKEN}" http://localhost:8092/updateuser/scim/Users/29db2ac2-3863-491d-9b2a-c23863491d75 -d'
   { "schemas":
       ["urn:ietf:params:scim:api:messages:2.0:PatchOp"],
     "Operations":[
       {
        "op":"remove",
        "path":"phoneNumbers"
       }
     ]
   }'
```

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/test/java/io/gravitee/am/gateway/handler/common/auth/UserAuthenticationServiceTest.java [hunk 1]
```diff
@@ -264,23 +263,56 @@ public void shouldConnect_knownUser_withRoles() {
...
         testObserver.assertNoErrors();
-        testObserver.assertValue(user1 -> user1.getRoles().size() == 2);
+        testObserver.assertValue(user1 -> user1.getDynamicRoles().size() == 2);
+        verify(userService).update(any(), argThat(actions -> actions.updateProfile()
+                && actions.updateDynamicRole()
+                && !(actions.updateAddresses() || actions.updateRole() || actions.updateEntitlements() || actions.updateAttributes())));
+    }
+
+    @Test
+    public void shouldConnect_user_withDynaicRoles_unchanged() {
```
Comment (a-cordier) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/test/java/io/gravitee/am/gateway/handler/common/auth/UserAuthenticationServiceTest.java:285]:
```suggestion
    public void shouldConnect_user_withDynamicRoles_unchanged() {
```

File: gravitee-am-repository/gravitee-am-repository-api/src/main/java/io/gravitee/am/repository/management/api/CommonUserRepository.java [hunk 2]
```diff
@@ -52,4 +55,102 @@ public interface CommonUserRepository extends CrudRepository<User, String> {
     Maybe<User> findById(ReferenceType referenceType, String referenceId, String userId);
 
     Completable deleteByReference(ReferenceType referenceType, String referenceId);
+
+    class UpdateActions {
+        private final boolean profile = true;
```
Comment (ashraf706) [inline gravitee-am-repository/gravitee-am-repository-api/src/main/java/io/gravitee/am/repository/management/api/CommonUserRepository.java:60]: perhaps profile is always true.

File: gravitee-am-repository/gravitee-am-repository-api/src/main/java/io/gravitee/am/repository/management/api/CommonUserRepository.java [hunk 3]
```diff
@@ -52,4 +55,102 @@ public interface CommonUserRepository extends CrudRepository<User, String> {
...
+        public boolean updateAddresses() {
+            return addresses;
+        }
+
+        public UpdateActions updateAddresses(boolean addresses) {
+            this.addresses = addresses;
+            return this;
+        }
+
+        public boolean notOnlyProfile() {
```
Comment (ashraf706) [inline gravitee-am-repository/gravitee-am-repository-api/src/main/java/io/gravitee/am/repository/management/api/CommonUserRepository.java:118]: This method name is bit difficult to understand what it does, though hard to suggest a better name

**Review by ashraf706 — APPROVED** [review]

File: gravitee-am-repository/gravitee-am-repository-api/src/main/java/io/gravitee/am/repository/management/api/CommonUserRepository.java (hunk 2 above)
Comment (leleueri) [reply gravitee-am-repository/gravitee-am-repository-api/src/main/java/io/gravitee/am/repository/management/api/CommonUserRepository.java]: removed

File: gravitee-am-repository/gravitee-am-repository-api/src/main/java/io/gravitee/am/repository/management/api/CommonUserRepository.java (hunk 3 above)
Comment (leleueri) [reply gravitee-am-repository/gravitee-am-repository-api/src/main/java/io/gravitee/am/repository/management/api/CommonUserRepository.java]: As I removed the profile boolean, I renamed this method to "updateRequire"

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcOrganizationUserRepository.java [hunk 4]
```diff
@@ -587,13 +599,29 @@ private Stream<JdbcOrganizationUser.Attribute> convertAttributes(User item, List
...
     }
 
-    private Mono<Integer> deleteChildEntities(String userId) {
-        Mono<Integer> deleteRoles = template.delete(JdbcOrganizationUser.Role.class).matching(Query.query(where(FK_USER_ID).is(userId))).all();
-        Mono<Integer> deleteRolesDynamicRoles = template.delete(JdbcOrganizationUser.DynamicRole.class).matching(Query.query(where(FK_USER_ID).is(userId))).all();
-        Mono<Integer> deleteAddresses = template.delete(JdbcOrganizationUser.Address.class).matching(Query.query(where(FK_USER_ID).is(userId))).all();
-        Mono<Integer> deleteAttributes = template.delete(JdbcOrganizationUser.Attribute.class).matching(Query.query(where(FK_USER_ID).is(userId))).all();
-        Mono<Integer> deleteEntitlements = template.delete(JdbcOrganizationUser.Entitlements.class).matching(Query.query(where(FK_USER_ID).is(userId))).all();
-        return deleteRoles.then(deleteRolesDynamicRoles).then(deleteAddresses).then(deleteAttributes).then(deleteEntitlements);
+    private Mono<Integer> deleteChildEntities(String userId, UpdateActions actions) {
```
Comment (remisultan) [inline gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcOrganizationUserRepository.java:602]: Let's make sure that for 4.0.0 and master we have `Mono<Long>` (due to the migration of the new bom dependencies)

**Review by remisultan — APPROVED** [review]
LGTM
a comment for future versions.

**Pull request merged**